python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,251 @@
+
+
+@to_bytes.register(IntegerType)
+@to_bytes.register(DateType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<i", value)
+
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
+@to_bytes.register(TimestamptzType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<q", value)
+
+
+@to_bytes.register(FloatType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<f", value)
+
+
+@to_bytes.register(DoubleType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<d", value)
+
+
+@to_bytes.register(StringType)
+def _(type_, value: Any) -> bytes:
+    return value.encode("UTF-8")
+
+
+@to_bytes.register(UUIDType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(type_, value: Any) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(type_, value: Any) -> bytes:
+    scale = abs(value.as_tuple().exponent)
+    quantized_value = value.quantize(Decimal("10") ** -scale)
+    unscaled_value = int((quantized_value * 10**scale).to_integral_value())
+    min_num_bytes = (unscaled_value.bit_length() + 7) // 8
+    return unscaled_value.to_bytes(min_num_bytes, "big", signed=True)
+
+
+@singledispatch
+def from_bytes(type_, b: bytes):
+    """A generic function which converts bytes to a built-in python value
+
+    Args:
+        type_(PrimitiveType): An implementation of the PrimitiveType base class
+        b(bytes): The bytes to convert
+    """
+    raise TypeError(f"Cannot deserialize bytes, type {type_} not supported: {str(b)}")
+
+
+@from_bytes.register(BooleanType)
+def _(type_, b: bytes):
+    return struct.unpack("<?", b)[0] != 0
+
+
+@from_bytes.register(IntegerType)
+@from_bytes.register(DateType)
+def _(type_, b: bytes):
+    return struct.unpack("<i", b)[0]
+
+
+@from_bytes.register(LongType)
+@from_bytes.register(TimeType)
+@from_bytes.register(TimestampType)
+@from_bytes.register(TimestamptzType)
+def _(type_, b: bytes):
+    return struct.unpack("<q", b)[0]
+
+
+@from_bytes.register(FloatType)
+def _(type_, b: bytes):
+    return struct.unpack("<f", b)[0]
+
+
+@from_bytes.register(DoubleType)
+def _(type_, b: bytes):
+    return struct.unpack("<d", b)[0]
+
+
+@from_bytes.register(StringType)
+def _(type_, b: bytes):
+    return bytes(b).decode("utf-8")
+
+
+@from_bytes.register(UUIDType)
+def _(type_, b: bytes):
+    return uuid.UUID(int=struct.unpack(">QQ", b)[0] << 64 | struct.unpack(">QQ", b)[1])
+
+
+@from_bytes.register(BinaryType)
+@from_bytes.register(FixedType)
+def _(type_, b: bytes):
+    return b
+
+
+@from_bytes.register(DecimalType)
+def _(type_, b: bytes):
+    return Decimal(int.from_bytes(b, "big", signed=True) * 10**-type_.scale).quantize(
+        Decimal("." + "".join(["0" for i in range(1, type_.scale)]) + "1"), rounding=ROUND_HALF_UP
```

Comment:
Question: I saw in another PR some assertions on the python type major version (checking for 3).

Are we supporting Python 2.7? If so, do we need to be concerned about differences between `range` semantics between the two?

Reply:
No we're definitely not supporting Python 2 since it's been officially deprecated for quite some time. Currently we're aiming to support 3.7+.